clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
For 1), I am more inlined rename or create a new call in KafkaConsumer, say fetchRecords(). Because the user of poll() might also expect it to do other IO, e.g. heartbeating, auto-commits, instead of only fetching data.

For 2), I am reluctant to make changes that I am not strong with. Because I am concerned that, if I do this change, another reviewer will question it which I can not defend. Then I need to discard the change I have made, which happened once in this ticket. I would be more confident to make this change if 1 committer or 2 senior non-committers can support it.